File: go-multi-region-s3-crr-kms-cmk/multi-region-s3-crr-kms-cmk/internal/stacks/multi-region-s3-crr-kms-cmk-source-stack.go
package stacks

import (
	"multi-region-s3-crr-kms-cmk/pkg/source"

	"github.com/aws/aws-cdk-go/awscdk/v2"
	"github.com/aws/aws-cdk-go/awscdk/v2/awss3"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type MultiRegionS3CrrKmsCmkSourceStackProps struct {
	StackProps                  awscdk.StackProps
	TargetBucket                awss3.Bucket
	TargetKeyIdSsmParameterName string
	TargetRegion                *string
}

type MultiRegionS3CrrKmsCmkSourceStack struct {
	Stack awscdk.Stack
}

func NewMultiRegionS3CrrKmsCmkSourceStack(scope constructs.Construct, id *string, props *MultiRegionS3CrrKmsCmkSourceStackProps) MultiRegionS3CrrKmsCmkSourceStack {
	stack := awscdk.NewStack(scope, id, &props.StackProps)

	source.NewMultiRegionS3CrrKmsCmkSource(stack, jsii.String("MySource"), &source.MultiRegionS3CrrKmsCmkSourceProps{
		TargetBucket:                props.TargetBucket,
		TargetKeyIdSsmParameterName: props.TargetKeyIdSsmParameterName,
		TargetRegion:                props.TargetRegion,
	})

	return MultiRegionS3CrrKmsCmkSourceStack{
		Stack: stack,
	}
}

File: ts-multi-region-s3-crr-kms-cmk/multi-region-s3-crr-kms-cmk/lib/multi-region-s3-crr-kms-cmk-source-stack.ts
import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { MultiRegionS3CrrKmsCmkSource } from 'multi-region-s3-crr-kms-cmk-source';

interface S3StaticMultiRegionSourceStackProps extends cdk.StackProps {
    targetBucket: s3.Bucket,
    targetKeyIdSsmParameterName: string,
    targetRegion: string
}

export class MultiRegionS3CrrKmsCmkSourceStack extends cdk.Stack {
    public targetBucket: s3.Bucket;
    public targetKeyIdSsmParameterName: string;

    constructor(scope: Construct, id: string, props: S3StaticMultiRegionSourceStackProps) {
        super(scope, id, props);

        const mySourceConstruct = new MultiRegionS3CrrKmsCmkSource(this, 'MySource', {
            targetBucket: props.targetBucket,
            targetKeyIdSsmParameterName: props.targetKeyIdSsmParameterName,
            targetRegion: props.targetRegion
        });
    }
}